---
title: "Introduction to scDEpipelineR6"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{1. Introduction to scDEpipelineR6}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
options(rmarkdown.html_vignette.check_title = FALSE)
```

## Introduction

scDEpipelineR6 is an R package for single cell differential gene exprssion analysis on single cell 
RNA-sequencing data. Currently this package can ingest count matrices as rds files, matrix market files, and single cell experiment files. Cell-level metadata is also required in order to use the package.

## Example data
```{r load_data}

#You will need some single cell data to use in your analysis




```

## Usage

### Filtering
```{r setup, eval=FALSE}
library(scDEpipelineR6)

#Create your sce object
sce <- BiostatsSingleCell$new(count_data = my_count,
                              meta_data = my_meta_cell,
                              sampleId_col = "sample",
                              cluster_col = "cell_type",
                              treatment_col = "diagnosis")

#Apply a first round filter
sce$apply_filter(min.perc.cells.per.gene = 0.01)

#Set up a group contrast
sce$set_group_mode(cluster_of_interest = "my_cell_type", ref_group = "my_ref_group", alt_group = "my_alt_group")

#Set up a second round filter
sce_qc <- sce$apply_filter_contrasts_R6(min.cells.per.gene = 3, min.perc.cells.per.gene = 0.1, perc.cells.filter = T,
                                                        min.cells.per.gene.type = "or", cells.per.gene.filter = T,
                                                        perc.filter = F, perc.filter.type = "and", perc_threshold = 0.75,
                                                        min.ave.pseudo.bulk.cpm = 1, pseudo.bulk.cpm.filter = F, min.cells.per.subj = 5)
```

### DE analysis

```{r de_analysis, eval=FALSE}
library(scDEpipelineR6)

#Select some covariates to use in your DE analysis or simulation analysis
covars <- c()

#Run a DE analysis of your choosing on your filtered data
#Here we will run nebula (HL method)
nebula.HL_results = sce_qc$nebula_pipeline(covs = covars,method="HL")$res.tab


```

### Simulation

```{r simulation, eval=FALSE}

#Run a simulation run
#Here we will simulate 3 DE simulations for nebula (HL method)
de_result <- de_sim_v10(in_dir = "/home/mryals/scRNAseq/DERIVED_DATA",
           gene_info_file = "GSE122960_gene_info.csv",
           gene_name_column = "index",
           meta_file = "GSE122960_meta_subset_Disease_format.csv",
           sample_column = "orig.ident",
           cell_type_column = "integrated.data.active.ident",
           cell_barcode_column = "prefix.cell",
           contrast_column = "Disease",
           contrast_ref_group = "Donor",
           contrast_alt_group = "Fibro",
           count_file = "GSE122960_count_subset.mtx",
           cluster = "AT1",
           covars = c("Age","Sex"),
           simulation_type = "de",
           fc_estimate = TRUE,
           simulation_mode = 3,
           fc=1.5,
           pDE=0.15,
           nSim.subjects = 8,
           n_simulation=50,
           main_seed=2021,
           percentage.cell=1.00,
           minimum.cells.per.gene.type = "and",
           minimum.cells.per.subject = 5,
           sim_cores = 5,
           de_methods = c('nebula_HL'),
           sim_seed_mode = "seed_mode",
           simulation_seeds = c(451268, 783780, 709681)
           )

```
